log/record_test.go

```diff
@@ -103,3 +103,63 @@ func TestRecordAttributes(t *testing.T) {
 		}
 	})
 }
+
+func TestRecordAllocationLimits(t *testing.T) {
+	const runs = 5
+
+	// Assign testing results to external scope so the compiler doesn't
+	// optimize away the testing statements.
+	var (
+		tStamp time.Time
+		sev    log.Severity
+		text   string
+		body   log.Value
+		n      int
+		attr   log.KeyValue
+	)
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.SetTimestamp(y2k)
+		tStamp = r.Timestamp()
+	}), "Timestamp")
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.SetObservedTimestamp(y2k)
+		tStamp = r.ObservedTimestamp()
+	}), "ObservedTimestamp")
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.SetSeverity(log.SeverityDebug)
+		sev = r.Severity()
+	}), "Severity")
+
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.SetSeverityText("severity text")
+		text = r.SeverityText()
+	}), "SeverityText")
+
+	bodyVal := log.BoolValue(true)
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.SetBody(bodyVal)
+		body = r.Body()
+	}), "Body")
+
+	attrVal := []log.KeyValue{log.Bool("k", true), log.Int("i", 1)}
+	assert.Equal(t, 0.0, testing.AllocsPerRun(runs, func() {
+		var r log.Record
+		r.AddAttributes(attrVal...)
```

Comment:
The `attrVal` should be inlined to show that it does not cause an allocation.
More: https://github.com/open-telemetry/opentelemetry-go/pull/4958#discussion_r1498886804

Reply:
Should we include `log.Slice`, `log.Bytes`, and `log.Map` in the attributes to test if a user would pass those to us it will cause allocations?

I'm not sure where we stop if we want to test user behavior here instead of just the `AddAttributes` method.

Reply:
> Should we include log.Slice, log.Bytes, and log.Map in the attributes to test if a user would pass those to us it will cause allocations?

Personally, I would include it as well. By the way, I think it will be better to add additional tests to address my concern. I will try to propose a PR tomorrow.

> I'm not sure where we stop if we want to test user behavior here instead of just the AddAttributes method.

The main reason why I am proposing it that this is also how prototype bridges are implemented:

- slog: https://github.com/open-telemetry/opentelemetry-go/blob/147762ffb5c4394417b3f973274cc570723db79b/log/internal/slog.go#L31-L34
- logr: https://github.com/open-telemetry/opentelemetry-go/blob/147762ffb5c4394417b3f973274cc570723db79b/log/internal/logr.go#L43-L56